Adds an ingress to redirect from Hub Preview to Hub

PuneetPunamiya opened the pull request:
- This patch adds an annotation which redirects from Hub Preview to Hub as
    `nginx.ingress.kubernetes.io/permanent-redirect: https://hub.tekton.dev`

Signed-off-by: Puneet Punamiya <ppunamiy@redhat.com>

<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/main/standards.md#principles) (if functionality changed/added)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/main/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/main/CONTRIBUTING.md) for more details._

piyush-garg commented:
/cc @afrittoli Will be great if you can review

afrittoli commented on config/04-kubernetes/42-hub-preview-redirect-ingress.yaml:
```diff
@@ -0,0 +1,21 @@
+    - secretName: ui-hub-preview-tekton-dev-tls
+      hosts: [hub-preview.tekton.dev]
+  rules:
+    - http:
```
I usually include `host: hub-preview.tekton.dev` as key under `rules` too, but I'm not 100% sure if it's required:

```
spec:
  rules:
  - host: hub-preview.tekton.dev
     http: (...)
```

PuneetPunamiya replied:
Sorry for late reply 
Is it fine if we apply the ingress on dogfooding first and see and if that works fine then merge the pr ??

afrittoli commented on config/04-kubernetes/42-hub-preview-redirect-ingress.yaml:
```diff
@@ -0,0 +1,21 @@
+  tls:
+    - secretName: ui-hub-preview-tekton-dev-tls
+      hosts: [hub-preview.tekton.dev]
+  rules:
```
I'm not sure this section is needed at all since we will redirect to hub in all cases, but I don't know if the ingress will be happy without it

afrittoli reviewed (commented):
Thank you for this!
Could you include the annotation for the DNS name too?
```
    dns.gardener.cloud/dnsnames: hub-preview.tekton.dev
    dns.gardener.cloud/ttl: "3600"
```

I guess it's ok for now to have this in the `hub` repo but I'd really prefer eventually to separate this out to the `plumbing` repo - because this is specific to the deployment to dogfooding and anyone installing `hub` does not really need this ingress at all.

afrittoli commented:
I tried on the cluster but unfortunately it doesn't work, we'd have to install a separate nginx controller. Using a CNAME record works, as long as the hub ingress is updated to include `hub-preview.tekton.dev` in the list of hosts.
In that case however there is no HTTP redirect sent to the client, simply the hub production site is served to `hub-preview` as well

PuneetPunamiya commented:
>   dns.gardener.cloud/dnsnames: hub-preview.tekton.dev
    dns.gardener.cloud/ttl: "3600"

Hey @afrittoli  I have added this annotation and updated the pr 

> I tried on the cluster but unfortunately it doesn't work, we'd have to install a separate nginx controller. Using a CNAME record works, as long as the hub ingress is updated to include hub-preview.tekton.dev in the list of hosts.
In that case however there is no HTTP redirect sent to the client, simply the hub production site is served to hub-preview as well

Should we try this on dogfooding once ?? Is there anything more we need to do from Hub Side ??

afrittoli commented:
I deployed the nginx ingress controller to dogfooding to watch the `tekton-hub-preview` namespace only:
```
helm repo add ingress-nginx https://kubernetes.github.io/ingress-nginx
helm repo update
helm install hub-preview-ingress-nginx ingress-nginx/ingress-nginx --set controller.watchNamespace=tekton-hub-preview --set controller.config.entries.default-server-return="301 https://hub.tekton.dev"
```

I also deployed this ingress resource, and the redirect works fine:
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: redirect-to-tekton-hub
  namespace: tekton-hub-preview
  annotations:
    acme.cert-manager.io/http01-edit-in-place: 'true'
    cert-manager.io/cluster-issuer: letsencrypt-prod
    nginx.ingress.kubernetes.io/permanent-redirect: https://hub.tekton.dev
    nginx.ingress.kubernetes.io/force-ssl-redirect: 'true'
    dns.gardener.cloud/dnsnames: hub-preview.tekton.dev
    dns.gardener.cloud/ttl: '3600'
    kubernetes.io/ingress.class: "nginx"
spec:
  tls:
    - secretName: ui-hub-preview-tekton-dev-tls
      hosts: [hub-preview.tekton.dev]
  rules:
    - host: hub-preview.tekton.dev
      http:
        paths:
          - path: /*
            backend:
              serviceName: ui
              servicePort: 8080
```

Could you update this PR to match the ingress above ^^^ ?

Thank you!

PuneetPunamiya commented:
> I deployed the nginx ingress controller to dogfooding to watch the `tekton-hub-preview` namespace only:
> 
> ```
> helm repo add ingress-nginx https://kubernetes.github.io/ingress-nginx
> helm repo update
> helm install hub-preview-ingress-nginx ingress-nginx/ingress-nginx --set controller.watchNamespace=tekton-hub-preview --set controller.config.entries.default-server-return="301 https://hub.tekton.dev"
> ```
> 
> I also deployed this ingress resource, and the redirect works fine:
> 
> ```
> apiVersion: extensions/v1beta1
> kind: Ingress
> metadata:
>   name: redirect-to-tekton-hub
>   namespace: tekton-hub-preview
>   annotations:
>     acme.cert-manager.io/http01-edit-in-place: 'true'
>     cert-manager.io/cluster-issuer: letsencrypt-prod
>     nginx.ingress.kubernetes.io/permanent-redirect: https://hub.tekton.dev
>     nginx.ingress.kubernetes.io/force-ssl-redirect: 'true'
>     dns.gardener.cloud/dnsnames: hub-preview.tekton.dev
>     dns.gardener.cloud/ttl: '3600'
>     kubernetes.io/ingress.class: "nginx"
> spec:
>   tls:
>     - secretName: ui-hub-preview-tekton-dev-tls
>       hosts: [hub-preview.tekton.dev]
>   rules:
>     - host: hub-preview.tekton.dev
>       http:
>         paths:
>           - path: /*
>             backend:
>               serviceName: ui
>               servicePort: 8080
> ```
> 
> Could you update this PR to match the ingress above ^^^ ?
> 
> Thank you!

@afrittoli  Thanks a lot for all your help, have updated the pr, could you please take a look at it once

afrittoli reviewed (commented):
Thank you for the updtes!
/lgtm

sm43 commented:
/approve
